Add library for reset interface

will-v-pi opened the pull request:
This separates the stdio_usb reset interface out into a separate library, which can be used by applications which use TinyUSB directly.

No change in behaviour when using the typical stdio_usb (eg in the hello_usb example).

Keeps the `PICO_STDIO_USB_` prefix on the config defines to maintain compatibility, and also maintains compatibility with picotool.

Instructions to use this are added to the header file.

See implementation in [debugprobe](https://github.com/raspberrypi/debugprobe/pull/178) and the [dev_multi_cdc example](https://github.com/raspberrypi/pico-examples/pull/694)

will-v-pi committed:
Allow use of reset interface with custom descriptors

will-v-pi committed:
Fix pico_config assertion error

will-v-pi committed:
Create separate library for reset interface

will-v-pi committed:
Move pico_usb_reset_interface library into separate folder

will-v-pi committed:
Fix bazel build

will-v-pi committed:
Move ms_os_20_desc into header

will-v-pi committed:
CFG_TUD_VENDOR isn't actually used

will-v-pi committed:
Add documentation to usb_reset_interface.h

will-v-pi committed:
Change #if PICO_ON_DEVICE -> #if LIB_PICO_USB_RESET_INTERFACE

Cleaner way of separating the usb_reset_interface.h header

kilograham commented on src/common/pico_usb_reset_interface_headers/include/pico/usb_reset_interface.h:
```diff
@@ -25,4 +41,128 @@
 // regular flash boot
 #define RESET_REQUEST_FLASH 0x02
 
+#if LIB_PICO_USB_RESET_INTERFACE
```
i would be tempted to move this into a separate header under pico_usb_reset_interface_headers, that can be included if desired

will-v-pi replied:
Have made a separate `usb_reset_interface_device.h` file which should be included instead of `usb_reset_interface.h`

kilograham commented on src/common/pico_usb_reset_interface_headers/include/pico/usb_reset_interface.h:
```diff
@@ -25,4 +41,128 @@
+#include "stdint.h"
+#include "device/usbd_pvt.h"
+
+void resetd_init(void);
```
not sure `resetd` is a sufficient/good prefix for the public API

kilograham replied:
note also, on this and the below, this is peculiar to TinyUSB - as we think about CherryUSB etc, we (i haven't looked in detail) should consider what if any of this is exposed (and whether we should expose them again in a separate header - e.g. one for TinyUSB one for CherryUSB)

will-v-pi replied:
I will rename them to `pico_usb_reset_interface_...`

Note that these functions are not supposed to be exposed as a public API - the only thing that should be used publicly is the `_resetd_driver` struct, I will try to refactor it so that's the case

will-v-pi replied:
I made a separate TinyUSB header, and made it so this should be compatible with the CherryUSB PR (as that only uses the `usb_reset_interface.h` header)

I couldn't refactor the functions out of the header, because I think the `pico_usb_reset_interface_driver` has to be defined in the header as `static const` in order for the debugprobe inclusion to compile, and therefore the functions must be in the header too

kilograham replied:
> I will rename them to `pico_usb_reset_interface_...`
> 
> Note that these functions are not supposed to be exposed as a public API - the only thing that should be used publicly is the `_resetd_driver` struct, I will try to refactor it so that's the case

note `pico_`/`hardware_` is not included in function names; e.g. `dma_foo`

will-v-pi committed:
Rename publically accessible functions

will-v-pi committed:
Move LIB_PICO_USB_RESET_INTERFACE only headers into separate files

will-v-pi committed:
Make new usb_reset_interface_device header so old usb_reset_interface header remains the same

lurch commented on src/common/pico_usb_reset_interface_headers/include/pico/usb_reset_interface_config.h:
```diff
@@ -0,0 +1,75 @@
+#endif
+#endif
+
+// PICO_CONFIG: PICO_STDIO_USB_RESET_INTERFACE_MS_OS_20_DESCRIPTOR_ITF, If vendor reset interface is included the USB interface number for the reset interface, type=int, default=2 if application is not using TinyUSB directly, undefined otherwise, group=pico_usb_reset_interface
```
```suggestion
// PICO_CONFIG: PICO_STDIO_USB_RESET_INTERFACE_MS_OS_20_DESCRIPTOR_ITF, If vendor reset interface is included this specifies the USB interface number for the reset interface, type=int, default=2 if application is not using TinyUSB directly, undefined otherwise, group=pico_usb_reset_interface
```
Perhaps this is slightly more readable?

lurch commented on src/common/pico_usb_reset_interface_headers/include/pico/usb_reset_interface_config.h:
```diff
@@ -0,0 +1,75 @@
+// PICO_CONFIG: PICO_STDIO_USB_RESET_INCLUDE_APP_DRIVER_CB, Set to 0 if your application defines usbd_app_driver_get_cb, type=bool, default=1 when using pico_usb_reset_interface, 0 otherwise, group=pico_usb_reset_interface
+#ifndef PICO_STDIO_USB_RESET_INCLUDE_APP_DRIVER_CB
+#define PICO_STDIO_USB_RESET_INCLUDE_APP_DRIVER_CB LIB_PICO_STDIO_USB
+#endif
```
```suggestion
// PICO_CONFIG: PICO_STDIO_USB_RESET_INCLUDE_DEFAULT_APP_DRIVER_CB, Set to 0 if your application defines its own usbd_app_driver_get_cb function, type=bool, default=1 when using pico_usb_reset_interface, 0 otherwise, group=pico_usb_reset_interface
#ifndef PICO_STDIO_USB_RESET_INCLUDE_DEFAULT_APP_DRIVER_CB
#define PICO_STDIO_USB_RESET_INCLUDE_DEFAULT_APP_DRIVER_CB LIB_PICO_STDIO_USB
#endif
```
Maybe? :man_shrugging:

lurch commented on src/common/pico_usb_reset_interface_headers/include/pico/usb_reset_interface_device.h:
```diff
@@ -0,0 +1,35 @@
+#define _PICO_USB_RESET_INTERFACE_DEVICE_H
+
+/** \file usb_reset_interface_device.h
+ *  \defgroup pico_usb_reset_interface pico_usb_reset_interface
```
Does this new `pico_usb_reset_interface` group need adding to `docs/index.h` ?

will-v-pi committed:
Review fixups

pico_usb_reset_... -> usb_reset_...

PICO_STDIO_USB_RESET_INCLUDE_APP_DRIVER_CB -> PICO_STDIO_USB_RESET_INCLUDE_DEFAULT_APP_DRIVER_CB

Add to docs/index.h

popcornmix commented:
Any reason this has stalled? I'd quite like to use this, so my project can handle reset to reconfigure in a new mode.

mtfurlan commented:
I've been using this PR in a project for a while now, which isn't great, but it has been working.

lurch commented:
Looks like this PR has conflicts which need resolving? @will-v-pi

will-v-pi commented:
> Looks like this PR has conflicts which need resolving? @will-v-pi

I've rebased it onto develop to resolve those - this PR is awaiting review by @kilograham

kilograham commented:
oops - sorry i seem to have ignored this!

1. we should probably remove the STDIO_ from the define names, and add backwards-compatibility in the least ugly place possible
2. we have hit a slight snafu obv with the fact that pico/usb_reset_interface (and particularly the header) existed before.. i'm kinda of loath to have a pico/foo where foo.h is not the API, so i wonder if we should rename this to something else

will-v-pi committed:
Merge remote-tracking branch 'origin/develop' into reset-lib

will-v-pi commented:
> oops - sorry i seem to have ignored this!
> 
> 1. we should probably remove the STDIO_ from the define names, and add backwards-compatibility in the least ugly place possible
> 2. we have hit a slight snafu obv with the fact that pico/usb_reset_interface (and particularly the header) existed before.. i'm kinda of loath to have a pico/foo where foo.h is not the API, so i wonder if we should rename this to something else

I can rename the new library to `pico_usb_reset_interface_device`, as the header name you actually include to use it is already `usb_reset_interface_device.h`. That will make it the longest library name, but it's only slightly longer than `pico_bootsel_via_double_reset` so it's probably fine. I can also move the new headers out of `common` and into that library.

will-v-pi committed:
Remove STDIO from defines

Include backwards compatibility with old define names

Also fix bazel build

will-v-pi commented:
Actually, maybe just `pico_usb_reset` is fine for the library name, given it also includes the resetting via baud rate

lurch commented on src/common/pico_usb_reset_interface_headers/include/pico/usb_reset_interface_config.h:
```diff
@@ -0,0 +1,132 @@
+#endif
+#endif
+
+// PICO_CONFIG: PICO_USB_RESET_INCLUDE_DEFAULT_APP_DRIVER_CB, Set to 0 if your application defines its own usbd_app_driver_get_cb function, type=bool, default=1 when using pico_usb_reset_interface, 0 otherwise, group=pico_usb_reset_interface
```
"default=1 when using pico_usb_reset_interface, 0 otherwise"
:confused: 

Search&replace typo?

will-v-pi replied:
```suggestion
// PICO_CONFIG: PICO_USB_RESET_INCLUDE_DEFAULT_APP_DRIVER_CB, Set to 0 if your application defines its own usbd_app_driver_get_cb function, type=bool, default=1 when using pico_stdio_usb, 0 otherwise, group=pico_usb_reset_interface
```

will-v-pi committed:
Merge pico_stdio_usb: fix overlap between activity-LED GPIO and active_low bit (#2947)

commit 97ed432d39da36146ea423ad7b0a175b3add1590
Author: Graham Sanderson <graham.sanderson@raspberrypi.com>
Date:   Tue May 19 17:37:03 2026 -0500

    grr

commit 90c5d5a023f5d0ae38efbb47b1a4b7efe289abb3
Author: Graham Sanderson <graham.sanderson@raspberrypi.com>
Date:   Tue May 19 17:36:27 2026 -0500

    missing ) in comment

commit 082e45980f52cfef103ac92700868d3bd8614020
Author: Graham Sanderson <graham.sanderson@raspberrypi.com>
Date:   Tue May 19 17:35:37 2026 -0500

    Re-arrange bit patterns to be more backwards compatible

commit e2436165ad8d1cd286afc577efbda62bf253febb
Author: Andrii Anoshyn <anoshyn.andrii@gmail.com>
Date:   Sat May 16 14:29:23 2026 +0300

    pico_stdio_usb: fix overlap between activity-LED GPIO and active_low bit

    The bootsel reset USB control transfer encodes the activity-LED GPIO in
    wValue. The current layout reads:

      if (request->wValue & 0x100) {
          gpio = request->wValue >> 9u;
      }
      active_low = request->wValue & 0x200;

    Bit 9 (0x200) is the active_low flag, but `gpio = wValue >> 9` also
    makes bit 9 the low bit of the extracted GPIO number. As a result:
    - Any odd-numbered activity-LED GPIO is silently flagged active_low.
    - Any client wanting active_low has the GPIO number's low bit forced.

    The active_low line was added in commit 5a988894 (\"add
    rom_reset_usb_boot_extra which supports >32 pins and ACTIVE_LOW\") on
    top of the original \"gpio = wValue >> 9\" encoding from commit 383e88ea,
    without shifting the gpio over by an extra bit to make room. See #2713.

    Shift the GPIO extraction to (wValue >> 10) so bit 9 is reserved for
    active_low alone. The resulting GPIO field is 6 bits (0-63), enough
    for RP2040 (30 GPIOs) and RP2350 (48 GPIOs). The canonical client
    (picotool) only sends wValue=disable_mask with bits 0-1 set, so this
    does not change any in-tree behavior; the change is only visible to
    out-of-tree clients that build wValue with the GPIO-specified bit set.

    Fixes #2713.

will-v-pi committed:
fix search & replace typo

lurch commented on src/rp2_common/pico_usb_reset/usb_reset.c:
```diff
@@ -0,0 +1,160 @@
+            bool active_low = false;
+#endif
+#if !PICO_USB_RESET_BOOTSEL_FIXED_ACTIVITY_LED
+            // wValue layout:
+            //   bits 0-1  : forwarded as the bootrom disable_interface_mask
+            //   bit 7     : 1 if the activity-LED GPIO is active-low
+            //   bit 8     : 1 if an activity-LED GPIO is being specified
+            //   bits 9-15 : activity-LED GPIO number (0-127)
+            if (request->wValue & 0x100) {
+                gpio = request->wValue >> 9u;
+                active_low = request->wValue & 0x80;
+            }
```
It _might_ be worth swapping around two of these lines, so that the code better matches the comment?
```suggestion
            // wValue layout:
            //   bits 0-1  : forwarded as the bootrom disable_interface_mask
            //   bit 7     : 1 if the activity-LED GPIO is active-low
            //   bit 8     : 1 if an activity-LED GPIO is being specified
            //   bits 9-15 : activity-LED GPIO number (0-127)
            if (request->wValue & 0x100) {
                active_low = request->wValue & 0x80;
                gpio = request->wValue >> 9u;
            }
```

will-v-pi replied:
I've swapped them, and replaced `0x100` etc with `(1u << 8)` to make that clearer as well

kilograham commented:
ok, so there was maybe a bit of a miscommunication on the renaming, but i have looked at it again anyway.

I am cool with leaving it called `pico_usb_reset_interface` since thats really what it is - the original headers can remain.

even though they are long, we should probably keep PICO_USB_RESET_INTERFACE_ in the names to be consistent (i could handle contraction to IFC i guess)

It's kinda ugly, but the real headers really belong in rp2_common, so i'd actually put them with the code in rp2_common, but add them to the pico_usb_reset_interface_headers, and #include them from the original one guarded by a `#if LIB_PICO_USB_RESET_INTERFACE`

will-v-pi commented:
> ok, so there was maybe a bit of a miscommunication on the renaming, but i have looked at it again anyway.
> 
> I am cool with leaving it called `pico_usb_reset_interface` since thats really what it is - the original headers can remain.
> 
> even though they are long, we should probably keep PICO_USB_RESET_INTERFACE_ in the names to be consistent (i could handle contraction to IFC i guess)
> 
> It's kinda ugly, but the real headers really belong in rp2_common, so i'd actually put them with the code in rp2_common, but add them to the pico_usb_reset_interface_headers, and #include them from the original one guarded by a `#if LIB_PICO_USB_RESET_INTERFACE`

Sorry, I was occupied with other stuff - I think the rename to `pico_usb_reset` makes all of this cleaner so I will do that, as it allows for a separate pico_usb_reset_headers library so everything can be in rp2_common without any hacks in the existing header

will-v-pi committed:
Rename library to pico_usb_reset, and move headers into rp2_common

will-v-pi committed:
Fix bazel build, and remove some unnecessary backwards compatibility for new defines

will-v-pi committed:
fix bazel formatting

lurch commented on src/rp2_common/pico_usb_reset/include/pico/usb_reset_config.h:
```diff
@@ -0,0 +1,126 @@
+#define PICO_USB_RESET_BOOTSEL_ACTIVITY_LED PICO_STDIO_USB_RESET_BOOTSEL_ACTIVITY_LED
+#endif // default is undefined
+
+// PICO_CONFIG: PICO_USB_RESET_BOOTSEL_ACTIVITY_LED_ACTIVE_LOW, Whether pin to use as bootloader activity LED when BOOTSEL mode is entered via USB (either VIA_BAUD_RATE or VIA_VENDOR_INTERFACE) is active low, type=bool, default=0, group=pico_usb_reset
```
https://github.com/raspberrypi/pico-sdk/pull/2971/files mentions that the activity led active_low setting is ignored on RP2040. Is the same true here? (And if so, should it be mentioned?)

lurch commented on src/rp2_common/pico_usb_reset/include/pico/usb_reset_config.h:
```diff
@@ -0,0 +1,126 @@
+#ifndef _PICO_USB_RESET_CONFIG_H
+#define _PICO_USB_RESET_CONFIG_H
+
+// PICO_CONFIG: PICO_USB_RESET_BOOTSEL_ACTIVITY_LED, Optionally define a pin to use as bootloader activity LED when BOOTSEL mode is entered via USB (either VIA_BAUD_RATE or VIA_VENDOR_INTERFACE), type=int, min=0, max=47 on RP2350B, 29 otherwise, group=pico_usb_reset
```
For readability purposes, I wonder if it might make sense to move all these `PICO_USB_RESET_BOOTSEL_` settings below `PICO_USB_RESET_SUPPORT_RESET_TO_BOOTSEL` ? (Since if the latter is set to 0, I guess none of the former settings are applicable?)

lurch commented on src/rp2_common/pico_usb_reset/include/pico/usb_reset_tusb.h:
```diff
@@ -0,0 +1,64 @@
+#if CFG_TUSB_DEBUG >= 2
+    .name = "RESET",
+#endif
+    .init             = usb_reset_interface_init,
+    .reset            = usb_reset_interface_reset,
+    .open             = usb_reset_interface_open,
+    .control_xfer_cb  = usb_reset_interface_control_xfer_cb,
+    .xfer_cb          = usb_reset_interface_xfer_cb,
```
To match the renaming done in a recent commit, would it also make sense to rename these functions to `usb_reset_init` etc.? Or does the `usb_reset_interface_` naming still make sense here? :shrug:

will-v-pi replied:
I think it still makes sense here, as these functions are specific to the USB interface that handles the reset

will-v-pi committed:
Reorder pico_config defines

Place vendor interface defines after PICO_ENABLE_USB_RESET_VIA_VENDOR_INTERFACE, and bootsel defines after PICO_USB_RESET_SUPPORT_RESET_TO_BOOTSEL

will-v-pi committed:
Mention that active_low is ignored on RP2040

will-v-pi committed:
Make wValue parsing better match the comment

lurch commented on src/rp2_common/pico_usb_reset/include/pico/usb_reset_tusb.h:
```diff
@@ -0,0 +1,64 @@
+  /* MS OS 2.0 Registry property descriptor: length, type */ \
+  U16_TO_U8S_LE(0x0080), U16_TO_U8S_LE(MS_OS_20_FEATURE_REG_PROPERTY), \
+  U16_TO_U8S_LE(0x0001), U16_TO_U8S_LE(0x0028), /* wPropertyDataType, wPropertyNameLength and PropertyName "DeviceInterfaceGUID" in UTF-16 */ \
+  'D', 0x00, 'e', 0x00, 'v', 0x00, 'i', 0x00, 'c', 0x00, 'e', 0x00, 'I', 0x00, 'n', 0x00, 't', 0x00, 'e', 0x00, \
+  'r', 0x00, 'f', 0x00, 'a', 0x00, 'c', 0x00, 'e', 0x00, 'G', 0x00, 'U', 0x00, 'I', 0x00, 'D', 0x00, 0x00, 0x00, \
```
Not directly-related to this PR (so feel free to ignore this suggestion), but _maybe_ this could be reformatted to:
```suggestion
  'D', 0,     'e', 0,     'v', 0,     'i', 0,     'c', 0,     'e', 0, \
  'I', 0,     'n', 0,     't', 0,     'e', 0,     'r', 0,     'f', 0, \
  'a', 0,     'c', 0,     'e', 0,     'G', 0,     'U', 0,     'I', 0, \
  'D', 0,     0, 0, \
```
as this would then match the formatting of the `{bc7398c1-...` UTF-16 string below?
